src/audio/smart_amp_maxim.c

```diff
@@ -0,0 +1,719 @@
+		 * re-send data in next resume when idle and the new
+		 * configuration will be used when playback/capture
+		 * starts.
+		 */
+		comp_err(dev, "smart_amp_ctrl_set_bin_data(): driver is busy");
+		return -EBUSY;
+	}
+
+	switch (cdata->data->type) {
+	case SOF_SMART_AMP_CONFIG:
+		ret = smart_amp_set_config(dev, cdata);
+		break;
+	case SOF_SMART_AMP_MODEL:
+		ret = comp_set_model(dev, &sad->model, cdata);
+		break;
+	default:
+		comp_err(dev, "smart_amp_ctrl_set_bin_data(): unknown binary data type");
+		break;
+	}
+
+	return ret;
+}
+
+static int smart_amp_ctrl_set_data(struct comp_dev *dev,
+				   struct sof_ipc_ctrl_data *cdata)
+{
+	int ret = 0;
+
+	/* Check version from ABI header */
+	if (SOF_ABI_VERSION_INCOMPATIBLE(SOF_ABI_VERSION, cdata->data->abi)) {
+		comp_err(dev, "smart_amp_ctrl_set_data(): invalid version");
+		return -EINVAL;
+	}
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_ENUM:
+		comp_dbg(dev, "smart_amp_ctrl_set_data(), SOF_CTRL_CMD_ENUM");
+		break;
+	case SOF_CTRL_CMD_BINARY:
+		comp_dbg(dev, "smart_amp_ctrl_set_data(), SOF_CTRL_CMD_BINARY");
+		ret = smart_amp_ctrl_set_bin_data(dev, cdata);
+		break;
+	default:
+		comp_err(dev, "smart_amp_ctrl_set_data(): invalid cdata->cmd");
+		ret = -EINVAL;
+		break;
+	}
+
+	return ret;
+}
+
+/* used to pass standard and bespoke commands (with data) to component */
+static int smart_amp_cmd(struct comp_dev *dev, int cmd, void *data,
+			 int max_data_size)
+{
+	struct sof_ipc_ctrl_data *cdata = data;
+
+	comp_dbg(dev, "smart_amp_cmd(): cmd: %d", cmd);
+
+	switch (cmd) {
+	case COMP_CMD_SET_DATA:
+		return smart_amp_ctrl_set_data(dev, cdata);
+	case COMP_CMD_GET_DATA:
+		return smart_amp_ctrl_get_data(dev, cdata, max_data_size);
+	default:
+		return -EINVAL;
+	}
+}
+
+static void smart_amp_free(struct comp_dev *dev)
+{
+	struct smart_amp_data *sad = comp_get_drvdata(dev);
+
+	comp_dbg(dev, "smart_amp_free()");
+
+	smart_amp_free_memory(sad, dev);
+
+	comp_free_model_data(dev, &sad->model);
+	rfree(sad);
+	rfree(dev);
+}
+
+static int smart_amp_verify_params(struct comp_dev *dev,
+				   struct sof_ipc_stream_params *params)
+{
+	int ret;
+
+	comp_dbg(dev, "smart_amp_verify_params()");
+
+	ret = comp_verify_params(dev, BUFF_PARAMS_CHANNELS, params);
+	if (ret < 0) {
+		comp_err(dev, "volume_verify_params() error: comp_verify_params() failed.");
+		return ret;
+	}
+
+	return 0;
+}
+
+static int smart_amp_params(struct comp_dev *dev,
+			    struct sof_ipc_stream_params *params)
+{
+	int err;
+
+	comp_dbg(dev, "smart_amp_params()");
+
+	err = smart_amp_verify_params(dev, params);
+	if (err < 0) {
+		comp_err(dev, "smart_amp_params(): pcm params verification failed.");
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int smart_amp_trigger(struct comp_dev *dev, int cmd)
+{
+	struct smart_amp_data *sad = comp_get_drvdata(dev);
+	int ret = 0;
+
+	comp_dbg(dev, "smart_amp_trigger(), command = %u", cmd);
+
+	ret = comp_set_state(dev, cmd);
+
+	if (ret == COMP_STATUS_STATE_ALREADY_SET)
```

Comment:
Is this actually heppening ? Pipeline should not be setting teh same state twice ?

Reply:
It was never happened on my test, but I'm not quite sure because several references are doing same thing.

[https://github.com/ryans-lee/sof/blob/sof_dsm_audio_component/src/audio/eq_fir/eq_fir.c#L748](url)
[https://github.com/ryans-lee/sof/blob/sof_dsm_audio_component/src/audio/volume/volume.c#L804](url)

Reply:
Looks okay without this condition check. Shall remove this.